### PR TITLE
Adding fine-tuning models to LUKE

### PR DESCRIPTION
**Pull request by ikuyamada**
# What does this PR do?

This pull request adds the following four fine-tuning models to LUKE:

* `LukeForMultipleChoice`
* `LukeForQuestionAnswering`
* `LukeForSequenceClassification`
* `LukeForTokenClassification`

LUKE was initially developed to solve entity-related NLP tasks, however, the model can also be used as a BERT-like pretrained model, thus has been frequently used to solve common NLP tasks such as text classification and question answering.
This pull request aims to enable users to easily solve such NLP tasks using LUKE.

Following BERT and RoBERTa, this pull request also adds the `classifier_dropout` property to `LukeConfig`.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@NielsRogge

### CODE REVIEW

File: tests/models/luke/test_modeling_luke.py [hunk 1]
```diff
@@ -139,15 +147,20 @@ def prepare_config_and_inputs(self):
             )
 
         sequence_labels = None
-        labels = None
+        token_labels = None
+        choice_labels = None
+        # labels = None
```
Comment (NielsRogge) [inline tests/models/luke/test_modeling_luke.py:152]: To be removed?

File: tests/models/luke/test_modeling_luke.py [hunk 2]
```diff
@@ -139,15 +147,20 @@ def prepare_config_and_inputs(self):
...
         entity_pair_classification_labels = None
         entity_span_classification_labels = None
 
         if self.use_labels:
             sequence_labels = ids_tensor([self.batch_size], self.type_sequence_label_size)
-            labels = ids_tensor([self.batch_size, self.seq_length], self.vocab_size)
+            token_labels = ids_tensor([self.batch_size, self.seq_length], self.num_labels)
+            choice_labels = ids_tensor([self.batch_size], self.num_choices)
+
+            # labels = ids_tensor([self.batch_size, self.seq_length], self.vocab_size)
```
Comment (NielsRogge) [inline tests/models/luke/test_modeling_luke.py:163]: To be removed?

**Review by NielsRogge — APPROVED** [review]
Thanks for adding this! Looks good to me.

File: src/transformers/models/luke/modeling_luke.py [hunk 3]
```diff
@@ -1635,3 +1778,456 @@ def forward(
...
+            total_loss = (start_loss + end_loss) / 2
+
+        if not return_dict:
+            output = (
+                start_logits,
+                end_logits,
+                outputs.hidden_states,
+                outputs.entity_hidden_states,
+                outputs.attentions,
+            )
```
Comment (sgugger) [inline src/transformers/models/luke/modeling_luke.py:2098]: Same here

File: src/transformers/models/luke/modeling_luke.py [hunk 4]
```diff
@@ -1635,3 +1778,456 @@ def forward(
...
+            loss_fct = CrossEntropyLoss()
+            loss = loss_fct(reshaped_logits, labels)
+
+        if not return_dict:
+            output = (
+                reshaped_logits,
+                outputs.hidden_states,
+                outputs.entity_hidden_states,
+                outputs.attentions,
+            )
```
Comment (sgugger) [inline src/transformers/models/luke/modeling_luke.py:2224]: Same here

File: src/transformers/models/luke/modeling_luke.py [hunk 5]
```diff
@@ -1635,3 +1778,456 @@ def forward(
...
+            position_ids=position_ids,
+            entity_ids=entity_ids,
+            entity_attention_mask=entity_attention_mask,
+            entity_token_type_ids=entity_token_type_ids,
+            entity_position_ids=entity_position_ids,
+            head_mask=head_mask,
+            inputs_embeds=inputs_embeds,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=True,
```
Comment (sgugger) [inline src/transformers/models/luke/modeling_luke.py:1848]: This will break jit compilation, `return_dict` should be left at the value set, then you should adapt the code to get the pooled_output.

File: tests/models/luke/test_modeling_luke.py (hunk 1 above)
Comment (ikuyamada) [reply tests/models/luke/test_modeling_luke.py]: This line is removed in the latest commit

**Review by sgugger — COMMENTED** [review]
Thanks for adding those new task-specific models. Please look at the implementation of other such models in the library as your implementation contains a couple of issues:
- you force `return_dict=True` for the base model, but we have this flag because some optimization modules do not support dict outputs, the flag should be passed along and then the code should be adapted to deal with base model outputs that are either tuples or `ModelOutput`.
- when returning a tuple, only the non-None values should be kept

Then you are adding the models to the auto-mapping and use the standard examples in their docstring but this model expects a lot more inputs, so I'm not sure if this is something we want. For instance, the masked-lm model was not added to the AutoModelForMaskedLM API in the original PR because its labels are not the same as other masked LM models.

I've left comments for this on the SequenceClassification model, but they apply to all the models you are adding.

**Comment by ikuyamada** [comment]
@sgugger Thank you so much for your detailed comments!

> you force return_dict=True for the base model, but we have this flag because some optimization modules do not support dict outputs, the flag should be passed along and then the code should be adapted to deal with base model outputs that are either tuples or ModelOutput.

I understand that we should not modify the `return_dict` value in the task-specific classes. However, the existing task-specific classes of LUKE (i.e., `LukeForMaskedLM`, `LukeForEntityClassification`, `LukeForEntityPairClassification`, and `LukeForEntitySpanClassification`) are also implemented by specifying `return_dict=True` to the base class. 

If we fix the code of these existing classes in this pull request, there is one issue: the `entity_last_hidden_state`, which is used in the existing task-specific classes, is positioned after the `hidden_states` and `attentions` in the `BaseLukeModelOutput` and `BaseLukeModelOutputWithPooling`. Therefore, to identify the numerical index of `entity_last_hidden_state`, we need to write some code that checks whether `output_attentions` and `output_hidden_states` are activated. To deal with this, I think it might be a good idea to change the position of the `entity_last_hidden_state` before the optional hidden-states and attention fields, but it breaks backward compatibility. I would appreciate any advices or tips to address this.

The new task-specific classes proposed in this pull request use only `last_hidden_state` and `pooled_output`, so the problem above happens only if we fix the existing task-specific classes.

> This model uses a lot of new inputs. Are we sur the standard example which will be included in the docstring will actually work (e.g. are all those entity inputs completely optional?)

The entity input of LUKE is completely optional. The model can run with only word-based inputs without issues.

**Comment by sgugger** [comment]
Ah, I hadn't caught the base model was doing it, and I see why it's more practical with the entity hidden states. Let's leave it as is for now then, and we will fix that in the future if a user opens an issue and actually needs the specific parts of PyTorch `return_dict=False`.

**Comment by ikuyamada** [comment]
@sgugger Thanks! I've pushed the commit that filters out None entries from the return values when `return_dict` is set to `False`. Please let me know if there are any issues.
